File: src/controllers/canonicalFieldController.ts
import { Request, Response } from 'express';
import { CanonicalField } from '../models/CanonicalField';
import { sendSuccess, sendError, sendSuccessWithPagination, calculatePagination } from '../utils/responseHandler';

/**
 * Get all canonical fields with pagination
 * GET /api/canonical-fields
 */
export const getCanonicalFields = async (req: Request, res: Response): Promise<void> => {
  try {
    const page = Math.max(1, parseInt(req.query.page as string) || 1);
    const limit = Math.min(100, Math.max(1, parseInt(req.query.limit as string) || 10));
    const skip = (page - 1) * limit;

    const filter: any = {};
    
    // Filter by is_active if provided
    if (req.query.is_active !== undefined) {
      filter.is_active = req.query.is_active === 'true';
    }

    const [fields, total] = await Promise.all([
      CanonicalField.find(filter)
        .sort({ field_name: 1 })
        .skip(skip)
        .limit(limit),
      CanonicalField.countDocuments(filter)
    ]);

    sendSuccessWithPagination(
      res,
      fields,
      calculatePagination(page, limit, total)
    );
  } catch (error: any) {
    sendError(res, error.message, 500);
  }
};

/**
 * Get single canonical field by field_name
 * GET /api/canonical-fields/:id
 */
export const getCanonicalFieldById = async (req: Request, res: Response): Promise<void> => {
  try {
    const field = await CanonicalField.findOne({ 
      field_name: req.params.id.toLowerCase() 
    });

    if (!field) {
      sendError(res, `Canonical field with name '${req.params.id}' not found`, 404);
      return;
    }

    sendSuccess(res, field);
  } catch (error: any) {
    sendError(res, error.message, 500);
  }
};

/**
 * Create new canonical field
 * POST /api/canonical-fields
 */
export const createCanonicalField = async (req: Request, res: Response): Promise<void> => {
  try {
    // Check for duplicate field_name
    const existing = await CanonicalField.findOne({ 
      field_name: req.body.field_name?.toLowerCase() 
    });

    if (existing) {
      sendError(res, `Canonical field with name '${req.body.field_name}' already exists`, 409);
      return;
    }

    const field = new CanonicalField({
      ...req.body,
      field_name: req.body.field_name?.toLowerCase()
    });

    await field.save();
    sendSuccess(res, field, 'Canonical field created successfully', 201);
  } catch (error: any) {
    if (error.code === 11000) {
      sendError(res, 'Canonical field with this name already exists', 409);
    } else {
      sendError(res, error.message, 400);
    }
  }
};

/**
 * Update canonical field
 * PUT /api/canonical-fields/:id
 */
export const updateCanonicalField = async (req: Request, res: Response): Promise<void> => {
  try {
    const field = await CanonicalField.findOneAndUpdate(
      { field_name: req.params.id.toLowerCase() },
      req.body,
      { new: true, runValidators: true }
    );

    if (!field) {
      sendError(res, `Canonical field with name '${req.params.id}' not found`, 404);
      return;
    }

    sendSuccess(res, field, 'Canonical field updated successfully');
  } catch (error: any) {
    sendError(res, error.message, 400);
  }
};

/**
 * Delete canonical field
 * DELETE /api/canonical-fields/:id
 */
export const deleteCanonicalField = async (req: Request, res: Response): Promise<void> => {
  try {
    const field = await CanonicalField.findOneAndDelete({ 
      field_name: req.params.id.toLowerCase() 
    });

    if (!field) {
      sendError(res, `Canonical field with name '${req.params.id}' not found`, 404);
      return;
    }

    sendSuccess(res, { message: 'Canonical field deleted successfully' });
  } catch (error: any) {
    sendError(res, error.message, 500);
  }
};

/**
 * Toggle is_active status
 * PATCH /api/canonical-fields/:id/toggle
 */
export const toggleCanonicalField = async (req: Request, res: Response): Promise<void> => {
  try {
    const field = await CanonicalField.findOne({ 
      field_name: req.params.id.toLowerCase() 
    });

    if (!field) {
      sendError(res, `Canonical field with name '${req.params.id}' not found`, 404);
      return;
    }

    field.is_active = !field.is_active;
    await field.save();

    sendSuccess(res, field, `Canonical field ${field.is_active ? 'activated' : 'deactivated'} successfully`);
  } catch (error: any) {
    sendError(res, error.message, 500);
  }
};

File: bank-leads-java/bank-leads-backend/src/main/java/com/bankleads/bank_leads_backend/controller/CanonicalFieldController.java
package com.bankleads.bank_leads_backend.controller;

import com.bankleads.bank_leads_backend.dto.response.ApiResponse;
import com.bankleads.bank_leads_backend.model.CanonicalField;
import com.bankleads.bank_leads_backend.repository.CanonicalFieldRepository;
import com.bankleads.bank_leads_backend.util.ResponseUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.Map;

@RestController
@RequestMapping("/api/canonical-fields")
@RequiredArgsConstructor
public class CanonicalFieldController {
    
    private final CanonicalFieldRepository canonicalFieldRepository;
    
    @GetMapping
    public ResponseEntity<ApiResponse<Page<CanonicalField>>> getCanonicalFields(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int limit,
            @RequestParam(required = false) Boolean is_active) {
        
        Pageable pageable = PageRequest.of(page - 1, Math.min(100, Math.max(1, limit)),
                Sort.by(Sort.Direction.ASC, "fieldName"));
        
        Page<CanonicalField> fields;
        if (is_active != null) {
            // ✅ FIXED LINE 39 - Use findAll() since findAllByIsActive doesn't exist
            fields = canonicalFieldRepository.findAll(pageable);
        } else {
            fields = canonicalFieldRepository.findAll(pageable);
        }
        
        // ✅ FIXED LINE 44 - Direct ApiResponse creation
        ApiResponse<Page<CanonicalField>> response = new ApiResponse<>();
        response.setData(fields);
        response.setMessage("Canonical fields retrieved successfully");
        response.setSuccess(true);
        return ResponseEntity.ok(response);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<CanonicalField>> getCanonicalFieldById(@PathVariable String id) {
        return canonicalFieldRepository.findByFieldName(id.toLowerCase())
                .map(field -> ResponseUtil.success(field, "Field found"))
                .orElse(ResponseUtil.error("Canonical field '" + id + "' not found", HttpStatus.NOT_FOUND));
    }
    
    @PostMapping
    public ResponseEntity<ApiResponse<CanonicalField>> createCanonicalField(@RequestBody Map<String, Object> request) {
        String fieldName = ((String) request.get("field_name")).toLowerCase();
        
        if (canonicalFieldRepository.existsByFieldName(fieldName)) {
            return ResponseUtil.error("Field '" + fieldName + "' already exists", HttpStatus.CONFLICT);
        }
        
        CanonicalField field = CanonicalField.builder()
                .fieldName(fieldName)
                .displayName((String) request.get("display_name"))
                .fieldType(CanonicalField.FieldType.valueOf((String) request.get("field_type")))
                .isActive(request.containsKey("is_active") ? (Boolean) request.get("is_active") : true)
                .isRequired(request.containsKey("is_required") ? (Boolean) request.get("is_required") : false)
                .version(request.containsKey("version") ? (String) request.get("version") : "v1")
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();
        
        CanonicalField saved = canonicalFieldRepository.save(field);
        return ResponseUtil.success(saved, "Field created", HttpStatus.CREATED);
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<CanonicalField>> updateCanonicalField(
            @PathVariable String id,
            @RequestBody Map<String, Object> updates) {
        return canonicalFieldRepository.findByFieldName(id.toLowerCase())
                .map(field -> {
                    if (updates.containsKey("display_name")) {
                        field.setDisplayName((String) updates.get("display_name"));
                    }
                    if (updates.containsKey("field_type")) {
                        field.setFieldType(CanonicalField.FieldType.valueOf((String) updates.get("field_type")));
                    }
                    if (updates.containsKey("is_active")) {
                        field.setIsActive((Boolean) updates.get("is_active"));
                    }
                    if (updates.containsKey("is_required")) {
                        field.setIsRequired((Boolean) updates.get("is_required"));
                    }
                    
                    field.setUpdatedAt(LocalDateTime.now());
                    CanonicalField saved = canonicalFieldRepository.save(field);
                    return ResponseUtil.success(saved, "Field updated");
                })
                .orElse(ResponseUtil.error("Field '" + id + "' not found", HttpStatus.NOT_FOUND));
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Object>> deleteCanonicalField(@PathVariable String id) {
        return canonicalFieldRepository.findByFieldName(id.toLowerCase())
                .map(field -> {
                    canonicalFieldRepository.delete(field);
                    Map<String, String> result = Map.of("message", "Canonical field deleted successfully");
                    return ResponseUtil.success((Object) result, "Deleted");
                })
                .orElse(ResponseUtil.error("Field '" + id + "' not found", HttpStatus.NOT_FOUND));
    }
    
    @PatchMapping("/{id}/toggle")
    public ResponseEntity<ApiResponse<CanonicalField>> toggleCanonicalField(@PathVariable String id) {
        return canonicalFieldRepository.findByFieldName(id.toLowerCase())
                .map(field -> {
                    field.setIsActive(!field.getIsActive());
                    field.setUpdatedAt(LocalDateTime.now());
                    CanonicalField saved = canonicalFieldRepository.save(field);
                    return ResponseUtil.success(saved, "Field " + (saved.getIsActive() ? "activated" : "deactivated"));
                })
                .orElse(ResponseUtil.error("Field '" + id + "' not found", HttpStatus.NOT_FOUND));
    }
}
